[DNM] drivers: spi: Add shim for nrfx SPI driver

anangl opened the pull request:
This adds a translation layer to make the nrfx driver for the legacy
(i.e. without EasyDMA) nRF SPI peripheral accessible via the updated
Zephyr's API of the SPI driver.
Slave mode (i.e. the SPIS peripheral) is not supported yet.

Board-specific configurations for a few nRF5 boards are also added
to the "spi_loopback" test.

Depends on #5921

Commit:
drivers/spi: Add support for TX or RX only modes on DW driver

As for RX only, computing the NDF will be used for EEPROM mode.
Only a way to determine EEPROM mode is missing.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Cleanup the Kconfig files

Split QMSI relevant part into its own file. Some config where using
prompt, some not: normalizing it by removing the prompt keyword where
relevant.

Reducing the file by using if/endif when relevant.

However, it still not fully clean default: cfg and default baudrate
should disappear. There is no default configuration to apply as long
as the controller is not configured to run from any part using spi API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Adapt Kconfig and generic context to enable slave support

Adding Kconfig options to set supported modes by the controller
(master, slave or both)

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Add slave mode support to the DesignWare driver

It involves a minor change on which register is configured.
Most of the change is with threshold handling.
Handling the Kconfig based supported mode per-port.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

locomuco committed:
drivers/ethernet: Switch enc28j60 to new SPI API

Let's use the new SPI API and ditch the old one.

Signed-off-by: Matthias Boesl <matthias.boesl@gmail.com>
Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/ieee802154: Switch CC2520 to new SPI API

Finally switching to new SPI API. It makes code simpler and performs
better.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/quark_se: Switch to native SPI DW driver

QMSI is not updated to latest SPI API. Switching to native DW driver,
until we find a way either to update the shim drivers in a
non-performant way, or updating QMSI itself enabling support for
scatter-gather type of buffers.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
boards/x86: Pinmux SPI port 2 relevantly on quark_se_c1000_devboard

This port is a slave SPI port only.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Make spi_config parameter constant

As the content of this struct will not be modified by drivers, it's
better to pass it as constant. Also, if someday struct device can be
made contant too, this change will make spi ready for registering the
spi_config into ROM directly.

Fixes #5839

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Add a dedicated Kconfig option for asynchronous mode enablement

Instead of using CONFIG_POLL, which is not directly related to SPI and
is a kernel option, let's have SPI_ASYNC instead. When enabled, it will
select POLL automatically.

Fixes #5839

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Removing eeprom mode in configuration

EEPROM mode can be, in fact, asserted from tx_bufs/rx_bufs thus removing
it.

Fixes #5839

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Reduce parameter number on transceive function

tx_bufs/tx_count and rx_bufs/rx_count can be hold in another dedicated
structure, thus reducing the number of parameters to transceive. This
permits to avoid using the stack when calling transceive.

Since we saved parameters, we can expose back the struct device pointer,
to stay consistent with other device driver APIs.

Fixes #5839

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/flash: Switch W25QXXDV driver to new SPI API

This saves quite some memory.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/sensors: Switch bme280 driver to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/sensors: Switch bmi160 driver to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/sensors: Switch adxl362 driver to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/sensors: Switch lis2dh driver to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/adc: Switch ti_adc108s102 driver to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/bluetooth: Switch SPI based HCI driver to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Disable legacy API by default

Let's start deprecation work of the SPI legacy API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
samples/drivers: Switch Fujistu FRAM sample to new SPI API

Replacing legacy API calls by news ones.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/bluetooth: Get rid completely of legacy SPI API in SPI HCI

No need to support legacy API there.
The special usage of CS in BlueNRG is centralized it dedicated functions
to avoid all the #ifdefs everywher.
Also managing transaction error.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Add octal MISO lines mode

Some controllers can provide up to 8 MISO lines.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Make cs attribute in struct spi_config constant

Content is never modified by any driver.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Add the possibility to request CS active high logic

Some device may need to be put up on CS high logic. The active low logic
is the default as usual, but it is now possible to request the active
high logic.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/drivers: Removing old SPI test

It was much more limited than the spi_loopback test, and of course using
the legacy API. Removing it then.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

MaureenHelm committed:
spi: Implement new spi api in the mcux dspi driver

Adds support for the new spi api to the mcux dspi shim driver. Does not
remove support for the legacy spi api since there are still consumers of
that api, particularly the mcr20a 802.15.4 driver.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
tests: spi_loopback: Add frdm_k64f configuration

Adds a board-specific configuration for the frdm_k64f to the
spi_loopback test.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

Commit:
api/spi: Precise a bit the documentation

Notify about helpers

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/quark_se: Enable SPI port 2 as a slave only

Quark SE provides a slave SPI.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Switch Intel driver to new SPI API

Ditch any legacy API support altogether.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Remove legacy DesignWare SPI driver

Nothing requires this driver anymore.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Specify options per-port on DW driver

Generalize clock control. Make interrupt policy per-port.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/clock_control: Enable ARC core support on quark_se driver

Clock control can also be done through the ARC core.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Enable port 3 and 4 on DW driver

This will be useful on Quark_SE ARC core which can access x86 core SPI
controller.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Fix typo on parameters type in DW arc regs definitions

s/u32_s/u32t obviously

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/quark_se_c1000_ss: Switch to SPI DW driver

As QMSI driver does not support new SPI API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/quark_d2000: Switch to SPI DW driver

As QMSI driver does not support new SPI API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Removing QMSI driver as it does not support new API

Native DW driver is relevantly used instead.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/spi: Remove excluded boards

These now support the new API through SPI DW native driver.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

anangl committed:
drivers: spi: Add shim for nrfx SPI driver

This adds a translation layer to make the nrfx driver for the legacy
(i.e. without EasyDMA) nRF SPI peripheral accessible via the updated
Zephyr's API of the SPI driver.
Configuration files are already prepared for adding support for SPIM
(Master with EasyDMA) and SPIS (Slave with EasyDMA) peripherals.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
tests: spi_loopback: Add configurations for a few nRF5 boards

This adds to the spi_loopback test board-specific configurations
for the following boards:
- nrf51_pca10028
- nrf52_pca10040
- nrf52840_pca10056

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

Commit:
drivers/ieee802154: Switch MCR20A driver to new SPI API

And adding support for GPIO CS as well.
It looks like the driver could benefit from centralizing all SPI access
into a unique function, the protocol does not seem too convoluted to do
so, like CC2520 or CC1200.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

mbolivar reviewed (commented):
I have tested this driver successfully with a modified samples/drivers/led_ws2812 on nrf52_pca10040 using SPI_0 on the Arduino headers.

I've got some coding style comments

anangl commented on drivers/spi/Kconfig.nrf5:
```diff
@@ -0,0 +1,217 @@
+
+if SPI_1_NRF5_SPI || SPI_1_NRF5_SPIM || SPI_1_NRF5_SPIS
+
+config SPI_1_NRF5
```
It should not, of course. This option was supposed to be replaced by `SPI_1_NRF5_SPI`. I somehow failed to remove this entry after inserting the `choice` section above.

anangl commented on drivers/spi/spi_nrf5.c:
```diff
@@ -0,0 +1,339 @@
+
+#include "spi_context.h"
+
+#ifdef __cplusplus
```
Right, it's completely unnecessary here. I'm surprised that I didn't notice its presence. It must have been present in the source file that I started with as a base and I got used to it so much that it just became transparent to me. Thank for pointing this out!

anangl commented on drivers/spi/spi_nrf5.c:
```diff
@@ -0,0 +1,339 @@
+	if (result != NRFX_SUCCESS) {
+		SYS_LOG_ERR("Failed to initialize device: %s",
+			    dev->config->name);
+		assert(false);
```
Yes, it should return an error. But this value is currently [ignored](https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/device.c#L54).
Added a `return` statement, and a comment.

tbursztyka replied:
Still: return a valid error code.

I have plans to check these and do something about it (I may introduce it in the relevant pr actually), so it's preferable that all drivers conforms to the signature, even if it's not handled right now. The day it will be: it will be seamless (no need to fix any drivers).

anangl commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -29,6 +29,7 @@ struct spi_cs_control spi_cs = {
 #else
 #define SPI_CS NULL
 #define CS_CTRL_GPIO_DRV_NAME ""
+
```
Right.

Commit:
drivers/spi: Remove legacy API support from mcux dspi driver

Now that MCR20A supports the new API, legacy support from mcux dspi
driver can be safely removed.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented on drivers/spi/spi_nrf5.c:
```diff
@@ -0,0 +1,339 @@
+				  get_nrf_spi_bit_order(spi_cfg->operation));
+	}
+
+	nrf_spi_frequency_set(spi->p_reg,
```
this should go into the if {} block above.

anangl replied:
I thought that frequency could be changed between `transceive` calls. Looks like I misunderstood the "rest of the attributes" in the note describing `spi_config`. I thought it referred to `operation` field only. I'll correct it.

anangl commented on drivers/spi/spi_nrf5.c:
```diff
@@ -0,0 +1,339 @@
+		return -EINVAL;
+	}
+
+	struct spi_context *ctx = &get_dev_data(spi_cfg->dev)->ctx;
```
Corrected.

anangl commented:
@mbolivar Thanks for looking into this. I addressed your comments.

anangl commented:
@tbursztyka Applied corrections.

tbursztyka commented:
@anangl I just put you patches in #5921

anangl commented:
@tbursztyka great!

Commit:
drivers/spi: Simplify how error is forwarded from ISR handler in DW

No need to test it twice then in completed().

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: spi_context lock makes transceive function reentrant in DW

No need to check on busy bit from hardware.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/ethernet: No need of semaphore for spi in enc28j60

SPI API is reentrant.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/ieee802154: No need of semaphore for spi in mcr20a

SPI API is reentrant.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
samples/bluetooth: Move hci_spi to new SPI API

Removing old API usage.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Remove legacy NRF5 master and slave drivers

As legacy SPI API is being removed.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
spi: Remove SPI legacy API

No drivers nor samples are using it anymore.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

anangl commented:
Merged via #5921.
